R/analytics_header.R

```diff
@@ -0,0 +1,285 @@
+GdprServer <- function(
+  id,
+  cookies,
+  google_analytics_id,
+  tab
+) {
+  shiny::moduleServer(
+    id,
+    function(input,
+             output,
+             session) {
+
+      # Scope magnittr pipe
+      `%>%` <- magrittr::`%>%`
+
+      # Get the parent session (to determine if running in HTTPS)
+      parent_session <- shiny::getDefaultReactiveDomain()$rootScope()
+
+      # Reactive Val to store whether or not the GDPR Notice has been accepted
+      # Defaults to False
+      is_analytics <- shiny::reactiveVal(FALSE)
+
+      # Run once when cookies.js has been loaded
+      shiny::observe({
+        # If the accept cookie is null show the GDPR notice
+        if (is.null(cookies()$accept)) {
+          GdprAlert()
+          # Else if the user has previously accepted the GDPR notice
+        } else if (as.logical(cookies()$accept)) {
+          # Debug Message
+          print("Previously Accepted")
+          # Include the analytics sctipt
+          output$analytics_script <- shiny::renderUI(
+            AddAnalytics(google_analytics_id)
+          )
+          # Set the Reacive Val to True (GDPR notice accepted)
+          is_analytics(TRUE)
+          # Else the user has not previously accepted the GDPR Notice
+        } else {
+          # Set the Reactive Val to False (GDPR notice declined)
+          is_analytics(FALSE)
+        }
+        # Bind once to cookies.js being initialised
+      }) %>% shiny::bindEvent(cookies(), once = TRUE)
+
+      # Observe the GDPR notice response
+      shiny::observe({
+        # Debug Message
+        print(input$cookie_accept)
+        # List for 'accept cookie' either
+        # TRUE if GDPR notice accepted or
+        # FALSE if GDPR notice declined
+        # N.B. this is an essential cookie
+        # so exempt from a seperate cookie notice
+        msg <- list(
+          name = "accept",
+          value = input$cookie_accept
+        )
+        # Check if running in HTTPS or HTTP
+        if (
+          shiny::isolate(
+            parent_session$clientData$url_protocol
+          ) == "https:"
+        ) {
+          # Debug Message
+          print("Running in HTTPS")
+          # If running in HTTPS set a secure cookie
+          session$sendCustomMessage("cookie-set-secure", msg)
+        } else {
+          # Debug Message
+          print("Running in HTTP")
+          # If running in HTTP set a 'normal cookie'
+          session$sendCustomMessage("cookie-set", msg)
+        }
+        # If the GDPR notice was accepted
+        if (as.logical(input$cookie_accept)) {
+          # Debug Message
+          print("Adding Analytics")
+          # Add the analytics script
+          output$analytics_script <- shiny::renderUI({
+            AddAnalytics(google_analytics_id)
+          })
+        }
+        # Set the Reactive Val to the response
+        is_analytics(as.logical(input$cookie_accept))
+        # Debug Message
+        print("Storing Cookie")
```

Comment:
Very minor, but could this be "Stored cookie", since it is printed after everything has been done?

Reply:
Changed.